web/modules/slavnews_weather/src/Plugin/Block/WeatherBlockFolder/WeatherBlock.php

```diff
@@ -23,39 +68,102 @@ class WeatherBlock extends BlockBase {
+  }
+
+  /**
+   * Receive all weather data to display.
+   */
+  public function getWeather() {
+    // Setting cid for cache that is saved.
+    $cacheId = "slavwnews_weather";
+    // Function that request's user geological location by ip.
+    $location = $this->getIplocation();
+
+    // If we have data in cache we do following.
+    if ($cache = $this->cacheBackend->get($cacheId)) {
+      // Get data from cache.
+      $data = $cache->data;
+      // Return data.
+      return [
+        '#theme' => 'slavnews__weather__block',
+        '#image' => $data['current']['condition']['icon'],
+        '#temp_c' => $data['current']['temp_c'],
+        '#city' => $data['location']['name'],
+      ];
+    }
+    // If we dont have data in cache we set it there.
+    // phpcs:ignore
+    elseif (\Drupal::config('AdminWeather.settings')->get('AdminWeather.settings.weatherapi_token') != NULL) {
+      // Get token from config file.
+      // phpcs:ignore
+      $weatherToken = \Drupal::config('AdminWeather.settings')->get('AdminWeather.settings.weatherapi_token');
+      // Full url.
+      $weather_url = 'http://api.weatherapi.com/v1/current.json?key=' . $weatherToken . '&q=' . $location . '&aqi=no';
+      // (pulling json data)
+      $response = $this->client->get($weather_url);
+      // Gettign all data from json.
+      $fullWeatherdata = json_decode($response->getBody($this->method), TRUE);
+      // Call function that caches.
+      $cachedData = $this->myCachingfunction($fullWeatherdata);
+
+      // After we get requested data we return it.
+      return $cachedData;
+    }
+    return [];
+  }
+
+  /**
+   * Validate and receive user IP.
+   */
+  public function getIplocation() {
+    // Request user Ip.
     // phpcs:ignore
-    $client = \Drupal::httpClient();
-    // (if it's our local ip, then we will hardcode location to Lutsk)
-    // @todo replace this static api validation on better one
+    $ip = \Drupal::request()->getClientIp();
+    $this->method = 'GET';
+    // If Ip is local we set default city.
     if (!filter_var($ip, FILTER_VALIDATE_IP, FILTER_FLAG_NO_PRIV_RANGE | FILTER_FLAG_NO_RES_RANGE)) {
       // (reverting to default if ip is local variable)
-      $ip_query = 'Lutsk';
+      // phpcs:ignore
+      return \Drupal::config('AdminWeather.settings')->get('AdminWeather.settings.admin_city') ?? NULL;
     }
+    // If Ip is not local do next if.
     else {
-      $ip_token = 'b0681a8c-ea7b-49c3-8dab-88900ca30bcb';
-      $ip_api_url = 'https://api.ipfind.com?ip=' . $ip . '&auth=' . $ip_token;
-      $response = $client->get($ip_api_url);
-      $ipData = json_decode($response->getBody($method), TRUE);
-      $ip_query = $ipData['country'] . ', ' . $ipData['city'];
+      // We request ipFind token from config, get ip and return...
+      // ready to use user geological location.
+      // phpcs:ignore
+      if ($ip_token = \Drupal::config('AdminWeather.settings')->get('AdminWeather.settings.ipfind_token') != NULL) {
```

Comment:
You have to use dependency injection here and below instead of calling static methods.